libsnark/polynomial_commitments/bdfg21.hpp

```diff
@@ -0,0 +1,139 @@
+ *             (originally developed by SCIPR Lab) and contributors
+ *             (see AUTHORS).
+ * @copyright  MIT license (see LICENSE file)
+ *****************************************************************************/
+
+#ifndef __LIBSNARK_POLYNOMIAL_COMMITMENTS_BDFG21_HPP__
+#define __LIBSNARK_POLYNOMIAL_COMMITMENTS_BDFG21_HPP__
+
+#include "libsnark/polynomial_commitments/kzg10.hpp"
+
+/// Reference:
+/// - [BDFG21]:
+///   Title: "Efficient polynomial commitment schemes for multiple points and
+///           polynomials"
+///   eprint: https://eprint.iacr.org/2020/081.pdf
+
+namespace libsnark
+{
+
+/// Batched polynomial commitment scheme from [BDFG21], using concepts from
+/// [KZG10] and [GWC19]. Supports evaluation of an arbitrary number of
+/// polynomials at arbitrary points, with a fixed-size witness. See [BDFG21]
+/// Section 4.
+///
+/// The srs and commitments are the same as KZG10.
+template<typename ppT> class bdfg21
+{
+public:
+    using Field = libff::Fr<ppT>;
+
+    /// The srs generated by the setup step. This is denoted PK in [KZG10]
+    /// section 3.2.
+    using srs = typename kzg10<ppT>::srs;
+
+    using commitment = typename kzg10<ppT>::commitment;
+
+    /// In [BDFG21], the evaluation points and polynomials are specified as:
+    ///
+    ///   - a set $\{ f_i \}$ of $k$ polynomials
+    ///
+    ///   - a set $T$ of all evaluation points
+    ///
+    ///   - the set $\{ S_i \}$ of subsets $S_i$ of $T$ where the polynomial
+    ///     $f_i$ is will be evaluated at the points in $S_i$.
+    ///
+    /// To match the expected use-cases, we assume that each polynomial $f_i$
+    /// is evaluated at exactly one point $z_i \in T$, and use a specification
+    /// which is efficient for this case.  Namely:
+    ///
+    ///   - `std::vector<Field> T` is a set of evaluation points.
+    ///
+    ///   - `std::vector<std::vector<polynomial>> f_sets` is a set of sets of
+    ///     polynomials, where `f_sets[i]` is the set of polynomials to be
+    ///     evaluated at `T[i]`.
+    ///
+    /// Repeated polynomials (that is, a polnomial `f` which is to be evaluated
+    /// at multiple points and therefore appears in multipled entries of
+    /// `f_sets`), would be very inefficient in this formulation, but are
+    /// expected to be rare.
+    ///
+    /// Other formulations can be supported as they are required.
+
+    /// The `i`-th entry is evaluations of polynomials in `f_sets[i]` at `z_i`.
+    using evaluations = std::vector<std::vector<Field>>;
+
+    using evaluation_witness_phase_1 = libff::G1<ppT>;
+
+    /// Contains the result of phase1 of the witness generation.
+    /// `public_witness_phase_1` is to be shared with the verifier, whereas
+    /// everything else is private context for use by the prover.
+    class phase_1_output
+    {
+    public:
+        libff::G1<ppT> public_witness_phase_1;
+        polynomial<Field> private_f_over_Z_T_polynomial;
```

Comment:
I mean, ideally this type should be defined in https://github.com/clearmatics/libfqfft/tree/master/libfqfft (and we should use it instead of `std::vector` everywhere), but this is out of scope of this PR. See: https://github.com/clearmatics/libfqfft/issues/5

Reply:
Yes, I considered this (as commented at the definition).  I also considered it better in libfqfft, so didn't move it in the end, but fine to move it out.  Should be moved in latest commit.